packages/suite/src/hooks/wallet/form/useUtxoSelection.ts

```diff
@@ -0,0 +1,103 @@
+import type { AccountUtxo, PROTO } from '@trezor/connect';
+
+type Props = UseFormMethods &
+    Pick<UseSendFormState, 'account' | 'composedLevels' | 'feeInfo'> & {
+        composeRequest: (field?: string) => void;
+    };
+
+export const useUtxoSelection = ({
+    account,
+    composedLevels,
+    composeRequest,
+    feeInfo,
+    register,
+    setValue,
+    watch,
+}: Props) => {
+    // register custom form field (without HTMLElement)
+    useEffect(() => {
+        register({ name: 'isCoinControl', type: 'custom' });
+        register({ name: 'selectedUtxos', type: 'custom' });
+    }, [register]);
+
+    const isCoinControl = watch('isCoinControl');
+    const selectedUtxos: AccountUtxo[] = watch('selectedUtxos') || [];
+    const selectedFee = watch('selectedFee');
+
+    const [spendableUtxos, dustUtxos]: [AccountUtxo[], AccountUtxo[]] = account.utxo
+        ? account.utxo.reduce(
+              ([previousSpendable, previousDust]: [AccountUtxo[], AccountUtxo[]], current) =>
+                  feeInfo.dustLimit && parseInt(current.amount, 10) >= feeInfo.dustLimit
+                      ? [[...previousSpendable, current], previousDust]
+                      : [previousSpendable, [...previousDust, current]],
+              [[], []],
+          )
+        : [[], []];
+
+    const allUtxosSelected =
+        !!selectedUtxos.length && selectedUtxos.length === spendableUtxos.length;
+    const composedLevel = composedLevels?.[selectedFee || 'normal'];
+
+    const composedInputs = useMemo(
+        () =>
+            composedLevel && 'transaction' in composedLevel ? composedLevel.transaction.inputs : [],
+        [composedLevel],
+    ) as PROTO.TxInputType[];
+    const preselectedUtxos = useMemo(
+        () =>
+            account.utxo?.filter(utxo =>
+                composedInputs.some(
+                    input => input.prev_hash === utxo.txid && input.prev_index === utxo.vout,
+                ),
+            ) || [],
+        [account.utxo, composedInputs],
+    );
+
+    const toggleCheckAllUtxos = () => {
+        if (allUtxosSelected) {
+            setValue('selectedUtxos', []);
+        } else {
+            setValue('selectedUtxos', spendableUtxos);
+            setValue('isCoinControl', true);
+        }
+        composeRequest();
+    };
+
+    const toggleCoinControl = () => {
+        setValue('isCoinControl', !isCoinControl);
+        setValue('selectedUtxos', isCoinControl ? [] : preselectedUtxos);
+        composeRequest();
+    };
+
+    const toggleUtxoSelection = (utxo: AccountUtxo) => {
+        const isSelected = selectedUtxos.find(u => u.txid === utxo.txid && u.vout === utxo.vout);
+        if (isSelected) {
+            setValue('selectedUtxos', selectedUtxos.filter(u => u !== isSelected) || []);
+        } else {
+            setValue(
+                'selectedUtxos',
+                !selectedUtxos.length && !isCoinControl
+                    ? [...new Set([...preselectedUtxos, utxo])]
```

Comment:
@dahaca please check 65b620e49e050a6fedd6e47a68d29cb9276f3454

Reply:
Honestly, I am not following what this code does, could you explain?

Reply:
Are `selectedUtxosOld` the pre-selected ones?

Reply:
`selectedUtxosOld` are the pre-selected ones only if Coin Control is not enabled and user turning the Coin Control on by selecting some UTXO. If Coin Control is already enabled, `selectedUtxosOld` are those already selected. I know it's a bit confusing, not sure how to name it better. Maybe it would be useful to add a comment?

<img width="787" alt="image" src="https://user-images.githubusercontent.com/3729633/188410332-90dee430-485d-4862-837d-77cc1722652e.png">

Reply:
Yes, a commend would be great. Also what about `preSelectedUtxos`?

Reply:
I added the comments and made a refactored it a little bit in 206f141ce511e95b25a7a076381ab768ffd3b1a8. I did not change the variable names because `preselectedUtxos` is already used.